rir/src/compiler/translations/pir_2_rir/allocators.h

```diff
@@ -0,0 +1,470 @@
+                    // falseBranch() next
+                    RegisterFile regC = reg;
+                    Stack stackC = stack;
+                    verifyBB(bb, bb->trueBranch(), regC, stackC);
+                }
+            }
+            if (bb->falseBranch() &&
+                !branchTaken.count(Jmp(bb, bb->falseBranch()))) {
+                branchTaken.insert(Jmp(bb, bb->falseBranch()));
+                verifyBB(bb, bb->falseBranch(), reg, stack);
+            }
+        };
+
+        {
+            RegisterFile f;
+            Stack s;
+            verifyBB(nullptr, code->entry, f, s);
+        }
+    }
+
+    size_t operator[](Value* v) const {
+        assert(allocation.at(v) != stackSlot);
+        return allocation.at(v) - 1;
+    }
+
+    size_t slots() const {
+        unsigned max = 0;
+        for (auto a : allocation) {
+            if (a.second != stackSlot && max < a.second)
+                max = a.second;
+        }
+        return max;
+    }
+
+    bool onStack(Value* v) const { return allocation.at(v) == stackSlot; }
+
+    bool hasSlot(Value* v) const { return allocation.count(v); }
+
+    size_t getStackOffset(Instruction* instr, std::vector<bool>& used,
+                          Value* what, bool remove) const {
+
+        auto stack = sa.stackBefore(instr);
+        assert(stack.size() == used.size());
+
+        size_t offset = 0;
+        size_t usedIdx = used.size() - 1;
+        auto i = stack.rbegin();
+        while (i != stack.rend()) {
+            if (*i == what) {
+                if (remove) {
+                    used[usedIdx] = true;
+                }
+                return offset;
+            }
+            if (hasSlot(*i) && onStack(*i) && !used[usedIdx]) {
+                ++offset;
+            }
+            ++i;
+            --usedIdx;
+        }
+        assert(false && "Value wasn't found on the stack.");
+        return -1;
+    }
+
+    size_t stackPhiOffset(Instruction* instr, Phi* phi) const {
+        auto stack = sa.stackBefore(instr);
+        size_t offset = 0;
+        for (auto i = stack.rbegin(); i != stack.rend(); ++i) {
+            if (*i == phi || phi->anyArg([&](Value* v) { return *i == v; }))
+                return offset;
+            if (hasSlot(*i) && onStack(*i))
+                ++offset;
+        }
+        assert(false && "Phi wasn't found on the stack.");
+        return -1;
+    }
+
+    // Check if v is needed after argument argNumber of instr
+    bool lastUse(Instruction* instr, size_t argNumber) const {
+        assert(argNumber < instr->nargs());
+        Value* v = instr->arg(argNumber).val();
+        auto stack = sa.stackAfter(instr);
+        for (auto i = stack.begin(); i != stack.end(); ++i)
+            if (*i == v)
+                return false;
+        for (size_t i = argNumber + 1; i < instr->nargs(); ++i)
+            if (instr->arg(i).val() == v)
+                return false;
+        return true;
+    }
+};
+
+struct nameAndEnvHash {
```

Comment:
this is very strange. on the one hand you call this nameAndEnvHash, suggesting that this is just for this name/env pair. but then the actual implementation is a template function, which is only necessary if you want to reuse this.

please just use `pairhash<SEXP,Value*>` from common.h